internal/security/secretstore/tokenfilewriter/tokenfilewriter_test.go

```diff
@@ -0,0 +1,189 @@
+
+	tests := []struct {
+		name                 string
+		rootToken            string
+		installPolicyCallErr error
+		createTokenCallErr   error
+	}{
+		{"Ok:CreateAndWrite with client call ok", testRootToken, nil, nil},
+		{"Bad:CreateAndWrite with empty token", "", nil, nil},
+		{"Bad:CreateAndWrite with install policy call error", testRootToken, testInstallPolicyErr, nil},
+		{"Bad:CreateAndWrite with create token call error", testRootToken, nil, testCreateTokenErr},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// setup mock secretclient and expected return values for this test
+			secretClient := &mocks.SecretStoreClient{}
+
+			secretClient.On("InstallPolicy", testRootToken,
+				consulSecretsEngineOpsPolicyName, mock.Anything).
+				Return(test.installPolicyCallErr).Once()
+
+			secretClient.On("CreateToken", testRootToken, mock.Anything).
+				Return(createTokenResult, test.createTokenCallErr).Once()
+
+			testTokenFilePath := filepath.Join(testTokenFileDir,
+				"mgmt-token.json"+"_"+strconv.FormatInt(time.Now().UnixNano(), 10))
+			fileWriter := NewWriter(lc, secretClient, flOpener)
+			_, err := fileWriter.CreateAndWrite(test.rootToken, testTokenFilePath,
+				fileWriter.CreateMgmtTokenForConsulSecretsEngine)
+			defer func() {
+				_ = os.RemoveAll(filepath.Dir(testTokenFilePath))
+			}()
+
+			expectError := test.installPolicyCallErr != nil ||
+				test.createTokenCallErr != nil ||
+				len(test.rootToken) == 0
+
+			if expectError {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				require.FileExists(t, testTokenFilePath)
+			}
+		})
+	}
+}
+
+func TestFileWriteErrorForConsulSecretEngine(t *testing.T) {
+	curUser, _ := user.Current()
+	if curUser != nil && curUser.Uid == "0" {
+		// it is root user then we skip this test as root can have permission to write everything
+		t.Skip()
+	}
+
+	lc, flOpener := setup()()
+
+	createTokenResult := getCreateTokenResultStub()
+
+	secretClient := &mocks.SecretStoreClient{}
+	testRootToken := "testRootToken"
+
+	secretClient.On("InstallPolicy", testRootToken,
+		consulSecretsEngineOpsPolicyName, mock.Anything).
+		Return(nil).Once()
+
+	secretClient.On("CreateToken", testRootToken, mock.Anything).
+		Return(createTokenResult, nil).Once()
+
+	// literally set to the /root/ directory so that there is no write permission
+	// as only the root user can perform this
+	testTokenFileDir := "/root/test"
+	testTokenFilePath := filepath.Join(testTokenFileDir, "mgmt-token.json")
+	fileWriter := NewWriter(lc, secretClient, flOpener)
+	_, err := fileWriter.CreateAndWrite(testRootToken, testTokenFilePath,
+		fileWriter.CreateMgmtTokenForConsulSecretsEngine)
+	require.Error(t, err)
+}
+
+func setup() func() (logger.LoggingClient, fileioperformer.FileIoPerformer) {
+	return func() (logger.LoggingClient, fileioperformer.FileIoPerformer) {
+		lc := logger.MockLogger{}
+		flOpener := fileioperformer.NewDefaultFileIoPerformer()
```

Comment:
Can you pass a mock object here instead of NewDefaultFileIoPerformer so that you can test file reading and writing without actually having to read and write files?

Reply:
i don't see a need here as there are already tests in`NewDefaultFileIoPerformer`  itself doing the mocking test why duplicate tests here?